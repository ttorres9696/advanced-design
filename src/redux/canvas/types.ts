import { Vector2d } from 'konva/types/types';

import { ShapeCoords } from '../../shared/components/ShapeElement/types';

export const SELECT_SHAPE = 'SELECT_SHAPE';
export const DESELECT_SHAPE = 'DESELECT_SHAPE';
export const INCREASE_ZOOM = 'INCREASE_ZOOM';
export const DECREASE_ZOOM = 'DECREASE_ZOOM';
export const NORMALIZE_ZOOM = 'NORMALIZE_ZOOM';
export const SET_CANVAS_ORIGIN = 'SET_CANVAS_ORIGIN';
export const INCREMENT_TOTAL_SHAPES = 'INCREMENT_TOTAL_SHAPES';
export const DECREMENT_TOTAL_SHAPES = 'DECREMENT_TOTAL_SHAPES';
export const SET_NEW_DRAGGABLE_POINT_MODE = 'SET_NEW_DRAGGABLE_POINT_MODE';
export const SET_MOUSE_POSITION = 'SET_MOUSE_POSITION';
export const SET_DELETE_DRAGGABLE_POINT_MODE = 'SET_DELETE_DRAGGABLE_POINT_MODE';
export const TOGGLE_MAGNET_MODE = 'TOGGLE_MAGNET_MODE';
export const SET_DRAGGING = 'SET_DRAGGING';
export const SET_SELECT_MODE = 'SET_SELECT_MODE';

export const originCoords = 20;
export type SelectionType = 'reset' | 'add';

interface SelectShapeAction {
  type: typeof SELECT_SHAPE;
  payload: {
    solarElementId: string;
    layer: string;
    selectionType: SelectionType;
  };
}

interface DeselectShapeAction {
  type: typeof DESELECT_SHAPE;
  payload: {
    solarElementId: string;
  };
}

interface IncreaseZoomAction {
  type: typeof INCREASE_ZOOM;
}

interface DecreaseZoomAction {
  type: typeof DECREASE_ZOOM;
}

interface NormalizeZoomAction {
  type: typeof NORMALIZE_ZOOM;
}

interface SetCanvasOriginAction {
  type: typeof SET_CANVAS_ORIGIN;
  payload: ShapeCoords;
}

interface IncrementTotalShapesAction {
  type: typeof INCREMENT_TOTAL_SHAPES;
}

interface DecrementTotalShapesAction {
  type: typeof DECREMENT_TOTAL_SHAPES;
  payload: number;
}

interface SetNewDraggablePointModeAction {
  type: typeof SET_NEW_DRAGGABLE_POINT_MODE;
  payload: boolean;
}

interface SetDeleteDraggablePointMode {
  type: typeof SET_DELETE_DRAGGABLE_POINT_MODE;
  payload: boolean;
}

interface SetMousePositionAction {
  type: typeof SET_MOUSE_POSITION;
  payload?: Vector2d;
}

interface ToggleMagnetModeAction {
  type: typeof TOGGLE_MAGNET_MODE;
}

interface SetDragging {
  type: typeof SET_DRAGGING;
  payload: boolean;
}

interface SetSelectMode {
  type: typeof SET_SELECT_MODE;
  payload: boolean;
}

export type CanvasActionTypes =
  | SelectShapeAction
  | DeselectShapeAction
  | IncreaseZoomAction
  | DecreaseZoomAction
  | NormalizeZoomAction
  | SetCanvasOriginAction
  | IncrementTotalShapesAction
  | DecrementTotalShapesAction
  | SetNewDraggablePointModeAction
  | SetDeleteDraggablePointMode
  | SetMousePositionAction
  | ToggleMagnetModeAction
  | SetDragging
  | SetSelectMode;
