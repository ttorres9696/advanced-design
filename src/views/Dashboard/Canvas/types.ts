import Konva from 'konva';
import { Node } from 'konva/types/Node';

import { SelectionType } from '../../../redux/canvas/types';
import { SolarDesign } from '../../../shared/models/SolarDesign.interface';

export interface CanvasProps {
  solarDesign: SolarDesign;
  canvasDimension: CanvasDimension;
}

export interface CanvasMargin {
  left?: number;
  top?: number;
  right?: number;
  bottom?: number;
}

export interface CanvasDimension {
  width: number;
  height: number;
}

export interface UseCanvasProps {
  solarDesign: SolarDesign;
  transformerSelectElement: (selectionType: SelectionType, node?: Node) => void;
  transformerHandleDragElement: (node: Node, e: Konva.KonvaEventObject<DragEvent>) => void;
  transformerOnDragEnd: (layer: string, node: Node) => string[];
}
