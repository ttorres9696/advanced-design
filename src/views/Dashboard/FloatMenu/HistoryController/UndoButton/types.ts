export interface UndoButtonProps {
  activeHistoryIndex: number;
}
