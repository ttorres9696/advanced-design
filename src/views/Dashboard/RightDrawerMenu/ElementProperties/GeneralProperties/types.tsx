export interface GeneralPropertiesProps {
  azimuth?: number;
  tilt?: number;
  updateProperty: (attr: string, value: any) => void;
  deleteObject: () => void;
}
