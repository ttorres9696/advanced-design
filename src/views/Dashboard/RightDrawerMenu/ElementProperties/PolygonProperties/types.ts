export interface PolygonPropertiesProps {
  points?: number[][];
  updateShapeProperty: (attr: string, value: any) => void;
  deletePoint: (index: number) => void;
}
