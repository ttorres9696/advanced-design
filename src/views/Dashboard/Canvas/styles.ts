import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';

export default makeStyles((theme: Theme) =>
  createStyles({
    content: {
      position: 'relative',
      zIndex: theme.zIndex.drawer + 2,
      width: '100%',
      '&:focus': {
        outline: 'none !important',
      },
    },
  }),
);
