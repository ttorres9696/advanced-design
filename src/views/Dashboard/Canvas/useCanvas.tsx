import Konva from 'konva';
import { Group } from 'konva/types/Group';
import { Layer } from 'konva/types/Layer';
import { KonvaEventObject, Node } from 'konva/types/Node';
import { Line } from 'konva/types/shapes/Line';
import { Stage } from 'konva/types/Stage';
import { Vector2d } from 'konva/types/types';
import { Transform } from 'konva/types/Util';
import { difference } from 'lodash';
import { useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { RootState } from '../../../redux';
import {
  incrementTotalShapes,
  selectShape,
  setCanvasOrigin,
  setDeleteDraggablePointMode,
  setMousePosition,
  setNewDraggablePointMode,
} from '../../../redux/canvas/actions';
import { SelectionType } from '../../../redux/canvas/types';
import { setDeleteDialogParams } from '../../../redux/core/actions';
import {
  addNewDraggablePoint,
  addShape,
  finishHistoryItem,
  setSelectedElement,
  setSelectedModuleSpec,
  startHistoryItem,
  toggleModule,
  updateShape,
} from '../../../redux/stage/actions';
import { ShapeCoords } from '../../../shared/components/ShapeElement/types';
import { TransformableElementChangeParams } from '../../../shared/components/TransformableElement/types';
import { getLowerCoords } from '../../../shared/geometry/SolarDesign';
import { getCurrentModulePointsRotation, getMagnetAlignment } from '../../../shared/helpers/modules.helper';
import { convertLinePointsToShapePoints } from '../../../shared/helpers/polygons.helper';
import { SolarElement } from '../../../shared/models/SolarElement.interface';
import { CanvasDimension, UseCanvasProps } from './types';

const useCanvas = ({
  solarDesign,
  transformerSelectElement,
  transformerHandleDragElement,
  transformerOnDragEnd,
}: UseCanvasProps) => {
  const dispatch = useDispatch();

  const [dragging, setDragging] = useState<boolean>(false);

  const layersOrder = useSelector((state: RootState) => state.stage.layersOrder);
  const layers = useSelector(
    (state: RootState) =>
      (state.stage.selectedModuleSpec &&
        state.stage.moduleSpecsStages[state.stage.selectedModuleSpec!.series].layers) || {
        trees: {},
        roofs: {},
        setbacks: {},
        modules: {},
        obstructions: {},
      },
  );

  const selectMode = useSelector((state: RootState) => state.canvas.selectMode);
  const selectedShapeElementIds = useSelector((state: RootState) => state.canvas.selectedShapeElementIds || []);
  const selectedLayerName = useSelector((state: RootState) => state.canvas.selectedLayerName || '');
  const magnetMode = useSelector((state: RootState) => state.canvas.magnetMode);
  const newDraggablePointMode = useSelector((state: RootState) => state.canvas.newDraggablePointMode);
  const mousePosition = useSelector((state: RootState) => state.canvas.mousePosition);
  const deleteDraggablePointMode = useSelector((state: RootState) => state.canvas.deleteDraggablePointMode);
  const canvasOrigin = useSelector((state: RootState) => state.canvas.canvasOrigin);
  const scale = useSelector((state: RootState) => state.canvas.scale);

  const [stageRef, setStageRef] = useState<Stage | null>(null);

  const onStartChange = useCallback((id: string, layer: string) => dispatch(startHistoryItem([id], layer)), [dispatch]);

  const onFinishChange = useCallback((id: string, layer: string) => dispatch(finishHistoryItem([id], layer)), [
    dispatch,
  ]);

  const selectElement = useCallback(
    (id: string, layer: string, selectionType: SelectionType, node?: Node) => {
      if (selectMode && (layer !== 'modules' || (layer === 'modules' && layers.modules[id].active))) {
        if (selectionType === 'reset') {
          selectedShapeElementIds.forEach((elementId: string) => {
            dispatch(setSelectedElement(selectedLayerName, elementId, false));
          });
        }

        if (id.length > 0 && (selectionType === 'reset' || (selectionType === 'add' && layer === selectedLayerName))) {
          dispatch(setSelectedElement(layer, id, true));
        }

        dispatch(selectShape(id, layer, selectionType));
        dispatch(setNewDraggablePointMode(false));
        dispatch(setDeleteDraggablePointMode(false));
        transformerSelectElement(selectionType, node);
      } else if (!selectMode && layer === 'modules') {
        onStartChange(id, layer);
        dispatch(toggleModule(id));
        onFinishChange(id, layer);
      }
    },
    [
      dispatch,
      transformerSelectElement,
      selectMode,
      selectedShapeElementIds,
      selectedLayerName,
      layers,
      onStartChange,
      onFinishChange,
    ],
  );

  const onRotateElement = useCallback(
    (id: string, layer: string, pointsRotationInRadians: number[], points: number[][]) =>
      dispatch(
        updateShape({
          ...layers[layer][id],
          shape: {
            ...layers[layer][id].shape,
            points,
            pointsRotationInRadians,
          },
        }),
      ),
    [dispatch, layers],
  );

  const onChangeElementParams = useCallback(
    (id: string, layer: string, params: TransformableElementChangeParams) => {
      if (params && params.points && params.points.length <= 2) {
        dispatch(
          setDeleteDialogParams({
            elementIds: [id],
            layer,
            open: true,
          }),
        );
      } else {
        dispatch(
          updateShape({
            ...layers[layer][id],
            shape: {
              ...layers[layer][id].shape,
              ...params,
            },
          }),
        );
      }
    },
    [dispatch, layers],
  );

  const onMoveStart = useCallback(
    (layer: string, node: Node) => {
      const elementIds: string[] = selectedShapeElementIds.includes(node.id())
        ? Array.from(selectedShapeElementIds)
        : [node.id()];

      dispatch(startHistoryItem(elementIds, layer));

      elementIds.forEach((elementId: string) => {
        dispatch(
          updateShape({
            ...layers[layer][elementId],
            shape: {
              ...layers[layer][elementId].shape,
            },
          }),
        );
      });
    },
    [dispatch, selectedShapeElementIds, layers],
  );

  const onMoveElement = useCallback(
    (id: string, layer: string, node: Node, e: Konva.KonvaEventObject<DragEvent>) => {
      transformerHandleDragElement(node, e);

      if (magnetMode && selectedShapeElementIds.length <= 1) {
        if (layer === 'modules') {
          const line = node as Line;
          const x = line.x();
          const y = line.y();

          const magnetAlignment = getMagnetAlignment(
            {
              ...layers[layer][id],
              shape: {
                ...layers[layer][id].shape,
                points: convertLinePointsToShapePoints(
                  line.points().map((coord: number, index: number) => {
                    if (index % 2 === 0) {
                      return coord + x;
                    } else {
                      return coord + y;
                    }
                  }) as number[],
                ),
              },
            },
            (Object.values(layers[layer]) as SolarElement[]).filter((element: SolarElement) => element.active),
            selectedShapeElementIds,
          );

          if (
            magnetAlignment &&
            difference(magnetAlignment.pointsRotationInRadians, layers[layer][id].shape.pointsRotationInRadians!)
              .length > 0
          ) {
            onChangeElementParams(id, layer, {
              ...magnetAlignment,
              points: magnetAlignment.points!.map(point => [point[0] - x, point[1] - y]),
            });
          }
        }
      }
    },
    [layers, onChangeElementParams, magnetMode, transformerHandleDragElement, selectedShapeElementIds],
  );

  const onMoveEnd = useCallback(
    (layer: string, node: Node) => {
      dispatch(finishHistoryItem(transformerOnDragEnd(layer, node), layer));
    },
    [dispatch, transformerOnDragEnd],
  );

  const onMouseMove = useCallback(
    (event: KonvaEventObject<MouseEvent>) => {
      if (selectedLayerName.length && newDraggablePointMode) {
        const stage: Stage = event.currentTarget as Stage;
        const masterLayer: Layer = stage
          .getLayers()
          .toArray()
          .find(layer => layer.attrs['id'] === 'layers') as Layer;

        if (masterLayer) {
          const selectedLayer: Group = masterLayer.findOne(`#${selectedLayerName}`) as Group;

          const transform: Transform = selectedLayer.getAbsoluteTransform().copy();
          transform.invert();

          const pointerPosition: Vector2d | null = stage.getPointerPosition();

          if (pointerPosition) {
            const relativePointerPosition: Vector2d = transform.point(pointerPosition);

            dispatch(setMousePosition(relativePointerPosition));
          }
        }
      } else if (mousePosition) {
        dispatch(setMousePosition(undefined));
      }
    },
    [selectedLayerName, newDraggablePointMode, dispatch, mousePosition],
  );

  const onMouseClick = useCallback(
    (event: KonvaEventObject<MouseEvent>) => {
      if (
        newDraggablePointMode &&
        mousePosition &&
        selectedLayerName.length > 0 &&
        selectedShapeElementIds.length === 1
      ) {
        dispatch(startHistoryItem(selectedShapeElementIds, selectedLayerName));

        dispatch(addNewDraggablePoint(selectedLayerName, selectedShapeElementIds[0], mousePosition));

        dispatch(finishHistoryItem(selectedShapeElementIds, selectedLayerName));
      }
    },
    [newDraggablePointMode, mousePosition, selectedLayerName, selectedShapeElementIds, dispatch],
  );

  useEffect(() => {
    dispatch(setSelectedModuleSpec(solarDesign.modules_spec));

    const lowerCoords: ShapeCoords = getLowerCoords(solarDesign, {
      top: 0,
      left: 150,
    });
    dispatch(setCanvasOrigin(lowerCoords));

    Object.values(solarDesign.layers)
      .reduce((allElements, layer) => {
        return allElements.concat(layer);
      }, [])
      .forEach((solarElement: SolarElement) => {
        if (solarElement.type === 'module') {
          solarElement.shape.pointsRotationInRadians = getCurrentModulePointsRotation(solarElement.shape.points!);
        }
        dispatch(addShape(solarDesign.modules_spec.series, solarElement));
        dispatch(incrementTotalShapes());
      });
  }, [solarDesign, dispatch]);

  useEffect(() => {
    if (stageRef) {
      const container = (stageRef as Stage).container();
      container.tabIndex = 1;
      container.focus();
      container.addEventListener(
        'keyup',
        (e: KeyboardEvent) => {
          if (selectedShapeElementIds.length > 0 && (e.key === 'Backspace' || e.key === 'Delete')) {
            dispatch(
              setDeleteDialogParams({
                elementIds: selectedShapeElementIds,
                layer: selectedLayerName,
                open: true,
              }),
            );
          }
        },
        false,
      );
    }
  }, [dispatch, selectedShapeElementIds, selectedLayerName, stageRef]);

  useEffect(() => {
    if (!selectMode && selectedShapeElementIds.length > 0) {
      selectedShapeElementIds.forEach((elementId: string) => {
        dispatch(setSelectedElement(selectedLayerName, elementId, false));
      });

      dispatch(selectShape('', '', 'reset'));
      dispatch(setNewDraggablePointMode(false));
      dispatch(setDeleteDraggablePointMode(false));
      transformerSelectElement('reset');
    }
  }, [dispatch, transformerSelectElement, selectMode, selectedShapeElementIds, selectedLayerName]);

  return {
    selectElement,
    selectedElementIds: selectedShapeElementIds,
    selectedLayerName,
    scale,
    canvasOrigin,
    onStartChange,
    onChangeElementParams,
    onRotateElement,
    onFinishChange,
    onMoveStart,
    onMoveElement,
    onMoveEnd,
    layersOrder,
    layers,
    newDraggablePointMode: newDraggablePointMode,
    deleteDraggablePointMode,
    mousePosition: mousePosition,
    selectMode: selectMode,
    onMouseMove,
    onMouseClick,
    setDragging,
    dragging,
    stageRef,
    setStageRef,
  };
};

export default useCanvas;
