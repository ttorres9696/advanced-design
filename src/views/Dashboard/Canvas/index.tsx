// eslint-disable-next-line
import Konva from 'konva';
import React from 'react';
import { Layer, Stage, Group, Transformer, Rect } from 'react-konva';

import { CanvasProps } from './types';
import useStyles from './styles';
import useCanvas from './useCanvas';
import { SolarElement } from '../../../shared/models/SolarElement.interface';
import TransformableElement from '../../../shared/components/TransformableElement';
import ShapeElement from '../../../shared/components/ShapeElement';
import NewDraggablePoint from '../../../shared/components/NewDraggablePoint';
import { availableTypesForResizing } from '../../../shared/helpers/polygons.helper';
import useTransformer from '../../../shared/hooks/useTransformer';
import { RotationArrowIconButtonProps } from '../../../shared/components/TransformableElement/RotationArrowIconButton/types';
import RotationArrowIconButton from '../../../shared/components/TransformableElement/RotationArrowIconButton';
import useRectSelection from '../../../shared/hooks/useRectSelection';

const Canvas: React.FC<CanvasProps> = ({ solarDesign, canvasDimension }) => {
  const classes = useStyles();

  const {
    transformerRef,
    transformerSelectElement,
    transformerRotationIcons,
    transformerHandleDragElement,
    transformerRotateElements90Degrees,
    transformerOnDragEnd,
  } = useTransformer();

  const {
    selectElement,
    selectedElementIds,
    selectedLayerName,
    onStartChange,
    onChangeElementParams,
    onRotateElement,
    onFinishChange,
    onMoveElement,
    onMoveStart,
    onMoveEnd,
    setStageRef,
    layersOrder,
    layers,
    scale,
    canvasOrigin,
    newDraggablePointMode,
    deleteDraggablePointMode,
    mousePosition,
    onMouseMove,
    onMouseClick,
    setDragging,
    dragging,
    selectMode,
  } = useCanvas({
    solarDesign,
    transformerSelectElement,
    transformerHandleDragElement,
    transformerOnDragEnd,
  });

  const {
    setRectSelectionRef,
    rectSelectionHandleTouchStart,
    rectSelectionHandleTouchMove,
    rectSelectionHandleTouchEnd,
  } = useRectSelection(selectElement);

  let activeLayersOrder: string[] = Array.from(layersOrder);
  if (selectedLayerName && selectedLayerName.length) {
    activeLayersOrder.splice(activeLayersOrder.indexOf(selectedLayerName), 1);
    activeLayersOrder.push(selectedLayerName);
  }

  if (!canvasDimension || !canvasDimension.width || !canvasDimension.height) {
    return null;
  }

  return (
    <Stage
      ref={ref => setStageRef(ref as any)}
      width={canvasDimension.width}
      height={canvasDimension.height}
      onClick={onMouseClick}
      className={classes.content}
      onMouseDown={e => {
        // deselect when clicked on empty area
        const clickedOnEmpty = e.target === e.target.getStage();
        if (clickedOnEmpty) {
          selectElement('', '', 'reset');
        }

        rectSelectionHandleTouchStart(e);
      }}
      onMouseMove={e => {
        onMouseMove(e);
        rectSelectionHandleTouchMove(e);
      }}
      onMouseUp={e => {
        rectSelectionHandleTouchEnd();
      }}
      onTouchStart={e => {
        console.log(`onTouchStart`);
        rectSelectionHandleTouchStart(e);
      }}
      onTouchMove={e => {
        console.log(`onTouchMove`);
        rectSelectionHandleTouchMove(e);
      }}
      onTouchEnd={e => {
        console.log(`onTouchEnd`);
        rectSelectionHandleTouchEnd();
      }}
      draggable={!selectMode}
      scaleX={scale}
      scaleY={scale}
      x={canvasOrigin.x}
      y={canvasOrigin.y}
    >
      <Layer id="layers">
        {layers
          ? activeLayersOrder.map(layerName => (
              <Group key={layerName} id={layerName}>
                {Object.keys(layers[layerName]).map(index => {
                  const solarElement: SolarElement = layers[layerName][index];

                  return (
                    <TransformableElement
                      key={solarElement.id}
                      id={solarElement.id}
                      layer={layerName}
                      shapeType={solarElement.shape.type}
                      version={solarElement.shape.version!}
                      isSelected={selectedElementIds.includes(solarElement.id)}
                      onSelect={selectElement}
                      onStartChange={onStartChange}
                      onChange={onChangeElementParams}
                      onRotate={onRotateElement}
                      onFinishChange={onFinishChange}
                      onMoveStart={onMoveStart}
                      onMove={onMoveElement}
                      onMoveEnd={onMoveEnd}
                      visible={solarElement.shape.visible!}
                      locked={solarElement.shape.locked!}
                      deleteDraggablePointMode={deleteDraggablePointMode}
                      resizeEnabled={availableTypesForResizing.includes(solarElement.type)}
                      setDragging={setDragging}
                      multiple={selectedElementIds.length > 1}
                      active={!!solarElement.active}
                      selectMode={selectMode}
                    >
                      <ShapeElement id={solarElement.id} {...solarElement.shape} solarType={solarElement.type} />
                    </TransformableElement>
                  );
                })}
              </Group>
            ))
          : null}

        <Transformer
          ref={transformerRef}
          resizeEnabled={false}
          rotateEnabled={selectedElementIds.length === 1}
          ignoreStroke={true}
          borderEnabled={true}
          borderDash={[5, 2]}
        >
          {selectedElementIds.length === 1
            ? transformerRotationIcons.map((icon: RotationArrowIconButtonProps) => (
                <RotationArrowIconButton
                  key={icon.position}
                  x={icon.x}
                  y={icon.y}
                  type={icon.type}
                  rotation={icon.rotation}
                  position={icon.position}
                  onClick={transformerRotateElements90Degrees}
                  invisible={dragging}
                />
              ))
            : null}
        </Transformer>

        <Rect ref={ref => setRectSelectionRef(ref)} fill="rgba(5, 102, 173, 0.5)" draggable={false} />
      </Layer>

      {newDraggablePointMode && mousePosition ? (
        <Layer id="new-draggable-point-mode">
          <NewDraggablePoint newDraggablePointMode={newDraggablePointMode} mousePosition={mousePosition} />
        </Layer>
      ) : null}
    </Stage>
  );
};

export default React.memo(Canvas);
