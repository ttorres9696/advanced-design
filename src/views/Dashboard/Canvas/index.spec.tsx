import { mount, ReactWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Provider } from 'react-redux';

import mockSolarDesign from '../../../redux/specs/mocks/solarDesign.mock';
import mockStore from '../../../redux/specs/mocks/state.mock';
import Canvas from '.';
import { CanvasProps } from './types';

let wrapper: ReactWrapper;

beforeEach(() => {
  const props: CanvasProps = {
    solarDesign: mockSolarDesign,
    canvasDimension: { width: 333, height: 222 },
  };

  wrapper = mount(
    <Provider store={mockStore}>
      <Canvas {...props} />
    </Provider>,
    {},
  );
});

describe('Canvas', () => {
  it('renders correctly', () => expect(toJson(wrapper)).toMatchSnapshot());
});
