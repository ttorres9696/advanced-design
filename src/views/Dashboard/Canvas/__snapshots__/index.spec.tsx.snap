// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardContent renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Memo(DashboardContent)
    isLeftDrawerOpened={true}
    isRightDrawerOpened={true}
  >
    <ForwardRef
      className="makeStyles-content-1 makeStyles-contentRightShift-2"
      draggable={true}
      height={336}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      scaleX={1}
      scaleY={1}
      width={160}
      x={0}
      y={0}
    >
      <Stage
        className="makeStyles-content-1 makeStyles-contentRightShift-2"
        draggable={true}
        forwardedRef={[Function]}
        height={336}
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseMove={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        scaleX={1}
        scaleY={1}
        width={160}
        x={0}
        y={0}
      >
        <div
          className="makeStyles-content-1 makeStyles-contentRightShift-2"
        />
      </Stage>
    </ForwardRef>
  </Memo(DashboardContent)>
</Provider>
`;
